ios/engine/KMEI/KeymanEngine/Classes/KeymanTextFieldDelegateProxy.swift

```diff
@@ -0,0 +1,73 @@
+//
+//  KeymanTextFieldDelegateProxy.swift
+//  KeymanEngine
+//
+//  Created by Gabriel Wong on 2017-10-06.
+//  Copyright © 2017 SIL International. All rights reserved.
+//
+//  Proxies delegate messages for a KeymanTextField.
+//    This allows the KeymanTextField to hook into these calls while allowing a developer to still
+//    use the delegate as normal (albeit with a different name: 'keymanDelegate')
+//
+//  This class is required because at the time of writing, setting a UITextField as it's
+//  own delegate caused an infinite loop in Apple's code, repeatedly calling the selectionDidChange callback
+//    See: http://lists.apple.com/archives/cocoa-dev/2010/Feb/msg01391.html
+
+import UIKit
+
+public protocol KeymanTextFieldDelegate: UITextFieldDelegate { }
+
+class KeymanTextFieldDelegateProxy: NSObject, UITextFieldDelegate {
+  weak var keymanDelegate: KeymanTextFieldDelegate?
+  private unowned let textField: UITextFieldDelegate
+
+  init(_ textField: KeymanTextField) {
+    self.textField = textField
+    super.init()
+  }
+
+  // MARK: - UITextFieldDelegate
+
+  // NOTE: Return values from the KeymanTextField hooks are ignored
+  func textFieldShouldBeginEditing(_ textField: UITextField) -> Bool {
+    _ = self.textField.textFieldShouldBeginEditing?(textField)
+    return keymanDelegate?.textFieldShouldBeginEditing?(textField) ?? true
```

Comment:
<3 that syntax.  So nice and compact compared to the original expanded if-else checks.